spec/fastsync/version2/fsv2_scheduler.tla

```diff
@@ -0,0 +1,532 @@
+  turn' = (
+    IF turn = "scheduler" THEN 
+      "environment" 
+    ELSE
+      "scheduler"
+  ) 
+
+Next == \* scheduler and environment alternate their steps (synchronous composition introduces more states)
+  /\ FlipTurn
+  /\ IF turn = "scheduler" THEN 
+       /\ NextSc
+       /\ inEvent' = noEvent
+       /\ UNCHANGED envRunning
+     ELSE
+       /\ NextEnv
+       /\ outEvent' = noEvent
+       /\ UNCHANGED <<scRunning, scheduler>>  
+
+(* ----------------------------------------------------------------------------------------------*)
+(* Expected properties                                                                           *)
+(* ----------------------------------------------------------------------------------------------*)
+
+TypeOK ==
+    /\ turn \in {"scheduler", "environment"}
+    /\ inEvent \in InEvents
+    /\ envRunning \in BOOLEAN
+    /\ outEvent \in OutEvents
+    /\ scheduler \in [
+         initHeight: Heights \cup {ultimateHeight + 1},
+         height: Heights \cup {ultimateHeight + 1},
+         peers: SUBSET PeerIDs,
+         peerHeights: [PeerIDs -> Heights \cup {None}],
+         peerStates: [PeerIDs -> PeerStates],
+         blocks: SUBSET Heights,
+         blockStates: [Heights -> BlockStates],
+         pendingBlocks: [Heights -> PeerIDs \cup {None}],
+         receivedBlocks: [Heights -> PeerIDs \cup {None}]
+       ]
+       
+TerminationWhenNoAdvance ==
+  (WF_turn(FlipTurn) /\ <> (inEvent.type = "tNoAdvanceExp"))
+    => <> ~scRunning
+
+NoFailuresAndTimeouts ==
+    \* no peer removal
+    /\ inEvent.type /= "peerRemoveEv"
+    \* no invalid blocks
+    /\ inEvent.type /= "bcNoBlockResponse"
+    /\ inEvent.type /= "pcBlockVerificationFailure"
+NoGlobalTimeout == inEvent.type /= "tNoAdvanceExp"
+
+FiniteResponse ==
+    <>[] (inEvent.type
+         \in {"bcAddNewPeer", "bcStatusResponse", "bcBlockResponse", "pcBlockProcessed", "pcBlockVerificationFailure", "rTrySchedule"})
+
+\* all blocks from initHeight up to max peer height have been processed
+AllRequiredBlocksProcessed ==
+  LET maxH == maxHeightScheduler(scheduler) IN
+  LET unprocessedBlocks == {\A h \in scheduler.initHeight.. maxH: scheduler.blockStates[h] /= "blockStateProcessed"} IN
+  scheduler.height >= maxH /\ Cardinality(unprocessedBlocks) = 0
+
+GoodTermination ==
+  (WF_turn(FlipTurn) /\ []NoGlobalTimeout /\ []NoFailuresAndTimeouts /\ FiniteResponse)
+     => <>(outEvent.type = "ScFinishedEv" /\ AllRequiredBlocksProcessed)
+
+\* TougherTermination is the termination property we like more and it holds true.
+\* When IncreaseHeight holds true, fastsync is progressing, unless it has hit the ceiling.
+IncreaseHeight ==
+    scheduler'.height > scheduler.height \/ scheduler'.height >= ultimateHeight
+
+TougherTermination ==
+  (scheduler.height < ultimateHeight
+    /\ (WF_turn(FlipTurn)
+    /\ SF_inEvent(OnTrySchedule)
+    /\ FiniteResponse
+    /\ (([]<> (<<IncreaseHeight>>_scheduler)) \/ <>(inEvent.type = "tNoAdvanceExp"))
```

Comment:
Can/should we add another property to ensure there are traces where this antecedent holds? Otherwise don't we risk implication based properties being trivial?

Reply:
Also why do we need `<<IncreaseHeight>>_scheduler`, and not just `IncreaseHeight` ?

Reply:
> Can/should we add another property to ensure there are traces where this antecedent holds? Otherwise don't we risk implication based properties being trivial?

I don't think I understand. Could you clarify?
> Also why do we need `<<IncreaseHeight>>_scheduler`, and not just `IncreaseHeight` ?

<> requires the state function, otherwise I get: `<> followed by action not of form <<A>>_v`

Reply:
eg. if FiniteResponse is always false then this property will be true.

Reply:
But isn't FiniteResponse an "eventually always true" temp. formula?...according to The Book (https://lamport.azurewebsites.net/pubs/lamport-spec-tla-plus.pdf, pg. 230-231):

line 524: `<>[]F asserts that eventually (at some time), F becomes true and remains true thereafter. In other words,<>[]F asserts that F is eventually always true.`

line 525: `[]<>F asserts that at all times, F is true then or at some later time.`

Will ask the experts tomorrow :)

Reply:
Yeah, there is always danger of writing a formula that is always false (unsatisfiable) in the left-hand side of implication. That makes the implication vacuously true. It is a good idea to check, whether the left-hand side is satisfiable.

Reply:
Igor and I spent some time on this and changed the properties and added proofs that preconditions are sometimes TRUE :)